tests/initcoder_test.go

```diff
@@ -0,0 +1,170 @@
+				"failed on transfer to empty receiver with valid code")
+
+			difference := receipt2.GasUsed - receipt1.GasUsed
+			require.Equal(difference, zeroByteCost,
+				"gas difference between 1 and 2 words should be 4, instead got", difference)
+		})
+
+		t.Run("aborts with init code size larger than MAX_INITCODE_SIZE", func(t *testing.T) {
+			// as specified in https://eips.ethereum.org/EIPS/eip-3860#rules,
+			// this is similar to transactions considered invalid for not meeting the intrinsic gas cost requirement.
+			_, err := runTransactionWithCodeSizeAndGas(t, net, MAX_INIT_CODE_SIZE+1, sufficientGas)
+			require.ErrorContains(err,
+				"intrinsic gas too low", "unexpectedly succeeded to create contract with init code length greater than MAX_INITCODE_SIZE")
+		})
+	})
+}
+
+func testForVariant(t *testing.T, net *IntegrationTestNet,
+	contract *initcode.Initcode, variant variant,
+	gasForContract, wordCost uint64) {
+
+	var gasCostFor2Words uint64 = wordCost * 3
+	// we use enough gas for all tests to afford cost of 3 words as well.
+	var gasFor3Words = gasForContract + gasCostFor2Words
+	t.Run("charges depending on the init code size", func(t *testing.T) {
+		// we need to provide same gas to all transactions so that we can compare the cost.
+		cost1Word, _ := createAndCheckContractWithCodeLenAndGas(t, net, contract, variant, 30, gasFor3Words)
+		cost2Words, _ := createAndCheckContractWithCodeLenAndGas(t, net, contract, variant, 42, gasFor3Words)
+		cost3Words, _ := createAndCheckContractWithCodeLenAndGas(t, net, contract, variant, 90, gasFor3Words)
+		require := require.New(t)
+		require.Equal(wordCost, cost2Words-cost1Word, "cost difference between 1 and 2")
+		require.Equal(wordCost*2, cost3Words-cost1Word, "cost difference between 1 and 3")
+	})
+
+	t.Run("fails without enough gas", func(t *testing.T) {
+		// 4 for a zero byte, 1 to make it fail.
+		receipt := createContractWithCodeLenAndGas(t, net, variant, 1, gasForContract-4-1)
+		require := require.New(t)
+		require.Equal(types.ReceiptStatusFailed, receipt.Status,
+			"unexpectedly succeeded to create contract without enough gas")
+	})
+
+	t.Run("with max init code size", func(t *testing.T) {
+		receipt := createContractWithCodeLenAndGas(t, net, variant, MAX_INIT_CODE_SIZE, sufficientGas)
+		require := require.New(t)
+		require.Equal(types.ReceiptStatusSuccessful, receipt.Status,
+			"failed to create contract with code length ", MAX_INIT_CODE_SIZE)
+	})
+
+	t.Run("aborts with init code size larger than MAX_INITCODE_SIZE", func(t *testing.T) {
+		receipt := createContractWithCodeLenAndGas(t, net, variant, MAX_INIT_CODE_SIZE+1, sufficientGas)
+		require := require.New(t)
+		require.Equal(types.ReceiptStatusFailed, receipt.Status,
+			"unexpectedly succeeded to create contract with init code length greater than MAX_INITCODE_SIZE")
+	})
+}
+
+func createAndCheckContractWithCodeLenAndGas(t *testing.T, net *IntegrationTestNet,
```

Comment:
nit: this function has too many responsibilities, and it feels difficult to read:
 I presume that this function was the result of a refactor that put together repeated code, nevertheless it lacks of a clear goal and responsibility. It just does too many things. 
 
I learned some rule, some long time ago: something like:  You do not do communication with the user inside the computation function, but instead return the result and write all IO (or stdout for the time being) at the same layer of the program. 

For this reason it would be better to have some pattern:
```
contract, err := createContract()
checkErr(err)
cost, err := getCost(contract)
checkErr(err)
```
Because in the domain of the unit test, error checking is reporting errors to the user. 
There are some reasons why this is this way, I will try to source some literature.

Reply:
I hope the new changes move towards what you comment here